contribute/engineering/backend/instrumentation.md

```diff
@@ -0,0 +1,207 @@
+# Instrumenting Grafana
+
+Guidance, conventions and best practices for instrumenting Grafana using logs, metrics and traces.
+
+## Logs
+
+Logs are files that record events, warnings and errors as they occur within a software environment. Most logs include contextual information, such as the time an event occurred and which user or endpoint was associated with it.
+
+Use the _pkg/infra/log_ package to create a named structured logger. Example:
+
+```go
+import (
+  "fmt"
+
+  "github.com/grafana/grafana/pkg/infra/log"
+)
+
+logger := log.New("my-logger")
+logger.Debug("Debug msg")
+logger.Warning("Warning msg")
+logger.Info("Info msg")
+logger.Error("Error msg", "error", fmt.Errorf("BOOM))
+```
+
+### Naming conventions
+
+Consider naming the logger using lowercase characters, e.g. `log.New("my-logger")` using snake_case or kebab-case styling.
+
+Consider to prefix the logger name with an area name if using different loggers across a feature/common packages, e.g. `log.New("plugin.loader")` and `log.New("plugin.client")`.
+
+Start the log message with a capital letter, e.g. `logger.Info("Hello world")` instead of `logger.Info("hello world")`.
+
+Use snake_case style when naming log keys, e.g. _remote_addr_ instead of _remoteAddr_.
```

Comment:
I disagree with this, and the codebase isn't consistent on this. We've for example consistently used `traceID`, including in the recent contextual logger refactoring.

https://github.com/grafana/grafana/blob/d9161c1a1bf9f1b19d9a1921f05982d175284f76/pkg/infra/tracing/tracing.go#L38

I think it makes sense to try to remain true to the Go identifier, which more likely will be camelCase than snake case, if anything, but I also think that being tolerant of _both_ isn't necessarily something that should be frowned upon here. It seems unnecessary to me to open up to nitpicking here.

Reply:
kind of agree - IMO we either use a linter for things like this or just write up a nice-to-have recommendation

Reply:
I thought people would have input on this :) Was looking at the HTTP logger middleware and it uses snake_case styling :) 
https://github.com/grafana/grafana/blob/48a8b35ece22fad106c94cbd03bc4de63c2ac886/pkg/middleware/logger.go#L57-L70

> I think it makes sense to try to remain true to the Go identifier, which more likely will be camelCase than snake case, if anything, but I also think that being tolerant of both isn't necessarily something that should be frowned upon here. It seems unnecessary to me to open up to nitpicking here.

Definitely sounds reasonable to me :+1: 

> just write up a nice-to-have recommendation

Yeah, I think I always considered camelCase for log keys, but after writing the conventions for metrics and traces I got the idea that using same style for everything would be consistent and allow switching from metrics (Prometheus) to logs (Loki) and possibly re-using the labels. But now I see that metric labels and log labels are not the same as log keys in general so I agree on camelCase.

```suggestion
Prefer using camelCase style when naming log keys, e.g. _remoteAddr_, to be consistent with Go identifiers. 
```